Packaged executable integration tests - initial integration tests with docker

tegefaulkes opened the pull request:
### Description

This PR modifies the `tests/bin` tests to target a packaged executable for the tests. This will be done by setting the `pk_pkg_executable_path` as the path to the desired executable we wish to test. This should then apply all bin tests to this executable.

The details are inside of issue MatrixAI/Polykey-CLI#10.

### Issues Fixed

* Fixes MatrixAI/Polykey#410 
* Fixes MatrixAI/Polykey#407 
* Fixes MatrixAI/Polykey#404
* Fixes MatrixAI/Polykey#420 <!-- hehe -->
* Fixes MatrixAI/Polykey#347
* Related MatrixAI/Polykey-CLI#10 
* Related MatrixAI/Polykey-CLI#12
* Related MatrixAI/Polykey-CLI#11 
* Related MatrixAI/Polykey-CLI#13 
* Related MatrixAI/Polykey-CLI#14

### Tasks
Given the scope of this PR, sub-tasks will be tracked with each issue.

- [x] 1. Allow bin tests to target the specified path when running the tests MatrixAI/Polykey#410 
- [x] 2. update tests to support docker image MatrixAI/Polykey#407 
- [x] 6. MatrixAI/Polykey#404 
- [x] 7. test optimisations MatrixAI/Polykey#420
- ~3. MatrixAI/Polykey-CLI#13~ - addressed in new PR
- ~4. update tests to support win build MatrixAI/Polykey-CLI#11~  - addressed in new PR
- ~5. update tests to support mac build MatrixAI/Polykey-CLI#12~  - addressed in new PR

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

tegefaulkes committed:
test: adding docker integration test

tegefaulkes committed:
feat: added bin utils functions for running target commands

This is to be used with the pkg integration tests to test the docker, windows and mac executables.

Related #410

tegefaulkes committed:
feat: added switching utility functions

These return either the pkX or the pkXTarget command if the `cmd` is set.

Related #410

tegefaulkes committed:
fix: updated `start.test.ts` to use #410 changes

tegefaulkes committed:
fix: updated `bootstrap.test.ts` to use #410 changes

tegefaulkes committed:
feat: adding tmdDir to globals

Allows us to override the tmp used for testing with `PK_TMP_DIR`.

tegefaulkes committed:
test: converting tests for docker target

- converting lock.test.ts
- cleaning up `testIf` commands and envs
- converted lockall.test.ts
- converted status.test.ts
- converted stop.test.ts
- converted unlock.test.ts

Related #407

tegefaulkes committed:
feat: added ability to override keypair generation with `privateKeyOverride` parameter for `createKeyManager`

This will skip key generation and use the provided `PrivateKey` instead. This should speed up testing by skipping the key generation.

Related #404

tegefaulkes committed:
feat: added `--private-key-file` option to `CommandStart.ts` and `CommandBootstrap.ts`

This should allow us to override the keypair generation with the provided private key. this will speed up agent starting.

Note that the key is provided as a Pem. The `PrivateKey` type contains functions that get destroyed somewhere between `commandStart.ts` and `keyManager.createKeyManager`. So I'm using the Pem string to keep the type primitive

Added `PK_ROOT_KEY` ENV and changed `--private-key-file` to `--root-key-file`

Related #404

tegefaulkes committed:
test: added `setupTestAgent` for creating test agents

This is a quick and easy way to create an agent with a pre-generated key.

Related #404

tegefaulkes committed:
test: removing global agent usage

I've also used the key override where needed for `createPolykeyAgent`.

Related #404

tegefaulkes committed:
fix: edge case bug for discovery queue

it wasn't handling the `ErrorConnectionNotRunning` that occurred in a very specific case.

tegefaulkes committed:
test: converted tests

Related #407

tegefaulkes committed:
test: overriding keypair generation for most tests

This should allow for faster testing since we skip the CPU intensive keypair generation.

Related #420

tegefaulkes committed:
tests: fixing tests for docker target

Related #407

tegefaulkes committed:
chore: updating the `.env.example` file

Related #407

tegefaulkes committed:
fix: tests use `global.tmpDir` directory

Also enabled `FF_NETWORK_PER_BUILD` flag for CI.

Related #407

tegefaulkes committed:
fix: adding `--rm` to `docker-run.sh` script.

This will remove the container when it ends.

#407

tegefaulkes committed:
tests: disabling tests

Disabling tests that don't work with docker for now.

#407

tegefaulkes committed:
ci: cleaning up CI jobs

CMCDragonkai commented on jest.config.js:
```diff
@@ -31,7 +31,8 @@ const globals = {
   testCmd: process.env.PK_TEST_COMMAND,
-  testPlatform: process.env.PK_TEST_COMMAND_DOCKER,
+  testPlatform: process.env.PK_TEST_PLATFORM ?? process.platform,
+  tmpDir: process.env.PK_TEST_TMP_DIR ?? os.tmpdir(),
```
The setting `tmpDir` here doesn't quite make sense to me.

Any code should be using `os.tmpdir()` should behave according to whatever `TMPDIR` is set to.

On Linux and containers, this is usually set to `/tmp`.

And all of our containers have `/tmp` built in.

Now if we want to "share" some data on the filesystem between container code and outside the container (whether that's a different container). Then the right way to do this is to:

1. Not use the `/tmp` for this.
2. Use a new path, make this `/srv`
3. Then mount a special path into `/srv`.

So for example when you are doing `docker run`.

You'd want to use:

```
docker run \
  --env PK_NODE_PATH='/srv' \
  --mount type=bind,src="$XXX",dst='/srv'
```

This is how we do this in the `testnet.polykey.io` task definition. Telling explicitly PK to use a particular node path, while mounting into the node path if we want to persist that data.

What is `$XXX`? This location is currently set to `PK_TEST_DATA_PATH`. This variable seems unnecessary. I think the `scripts/docker-run.sh` is intended to act as a "executable file proxy" so it can be used as a file target to execute. This variable should be passed in from the JavaScript code. It should not be hardcoded in `scripts/docker-run.sh`.

This means `tests/utils/exec.ts` (the location which should contain all the `pk*` functions, should decide where this location should be.

Now because you want to share the same location between the host (potentially a host container, and not host machine), and other containers. Then it should be sufficient to mount a temporary directory from `/tmp/...`.

CMCDragonkai replied:
Here's a visualisation (the 3 boxes are each separate environments, they may all be containers, or the host may be a machine/VM):

```
                                    ┌─────────────────────┐
                                    │    PolykeyAgent     │
                                    │                     │
                              ┌─────┼─►/srv               │
                              │     │  PK_NODE_PATH=/srv  │
                              │     └─────────────────────┘
                              │
┌────────────────────────┐    │
│ Host                   │    │
│                        │    │     ┌─────────────────────┐
│                        │    │     │   PolykeyClient     │
│ /tmp                   │    │     │                     │
│ /tmp/polykey-test-XXX──┼────┴─────┼─►/srv               │
│                        │          │  PK_NODE_PATH=/srv  │
└────────────────────────┘          └─────────────────────┘
```

The end result is that we do not change anything about the TMPDIR or `os.tmpdir`, and this variable should not be required.

This means the tests should look very similar as before, and you can use `--node-path` instead of `PK_NODE_PATH` to be more explicit.

CMCDragonkai replied:
If you already have a `pk*Target` let's just special case docker then.

Then you can specify the `--mount` option directly and also keep the paths the same.

I don't actually see a huge issue with using `/srv` as the directory node path inside the containers. But it can be set to the same path outside.

Therefore on CI/CD, we would see something like `/build/$CI_PROJECT_PATH/tmp/polykey-test-XXX` being used both inside the `PolykeyAgent` and `PolykeyClient` containers and as well as the same build/host container.

CMCDragonkai replied:
```
                                    ┌─────────────────────────┐
                                    │    PolykeyAgent         │
                                    │                         │
                              ┌─────┼─►/tmp/polykey-test-XXX  │
                              │     │  PK_NODE_PATH=""        │
                              │     └─────────────────────────┘
                              │
                              │
┌────────────────────────┐    │
│ Host                   │    │     ┌─────────────────────────┐
│                        │    │     │   PolykeyClient         │
│ /tmp                   │    │     │                         │
│ /tmp/polykey-test-XXX──┼────┴─────┼─►/tmp/polykey-test-XXX  │
│                        │          │  PK_NODE_PATH=""        │
└────────────────────────┘          └─────────────────────────┘
```

The `""` means the same path.

CMCDragonkai replied:
Alright so if `scripts/docker-run.sh` can be deleted and the same stuff done inside the `tests/utils/exec.sh`, we remove a couple environment variables.

Note however there is still these global/env variable:

1. `PK_TEST_PLATFORM` - defaults to `os.platform()` can be overridden to `'docker'`.
2. `PK_TEST_COMMAND` - target file to execute (it's supposed to be a shell command actually), in the case of `docker` execution, this could be set to the path to `docker` itself...
3. `PK_TEST_DATA_PATH` - the data path to be used by the tests, right now all the tests are using `os.tmpdir()`, it should default to using `PK_TEST_DATA_PATH`, which itself defaults to `os.tmpdir()`, on the CI/CD this is set to under the `/build/$CI_PROJECT_PATH` because only this location shareable between the the dind containers

CMCDragonkai replied:
Try and see what happens if you were to allow `shell` execution for `PK_TEST_COMMAND`. I believe we normally use `execFile`, but it can be changed to `exec`. Some things may not work. This will end up allowing arbitrary shell commands.

CMCDragonkai replied:
Ok `PK_TEST_COMMAND` is ignored when `PK_TEST_PLATFORM='docker'`.

tegefaulkes replied:
These changes have been made. Mainly the `docker-run.sh` has been removed and inlined within the `pkXTarget` functions.

CMCDragonkai commented on .gitlab-ci.yml:
```diff
@@ -321,16 +321,22 @@ integration:docker:
+        PK_TEST_DOCKER_IMAGE=$image \
+        PK_TEST_COMMAND=scripts/docker-run.sh \
+        PK_TEST_PLATFORM=docker \
+        PK_TEST_TMP_DIR=/builds/$CI_PROJECT_PATH/tmp \
```
So the directory here has to exist under `/builds`.

Any directory that is mounted into the Polykey containers subsequently have to be under this directory.

So the issue is that normally our test code would use `os.tmpdir()` to create the data directories.

But now this has to be changed to use this location instead.

To resolve this, the tests may require an override on the data directory creation, rather than creating from `os.tmpdir()`. But don't override `TMPDIR` env variable, that should be kept as is.

CMCDragonkai replied:
Therefore `PK_TEST_TMP_DIR` can actually be `PK_TEST_DATA_DIR`. Which would then override `os.tmpdir()`.

tegefaulkes replied:
`PK_TEST_TMP_DIR` is now `PK_TEST_DATA_DIR`.

tegefaulkes committed:
fix: removing `pkXSwitch` in favor of `pkX`

#410

tegefaulkes committed:
fix: moved `globalRootKeyPems.ts` to fixtures

#404
#420

tegefaulkes committed:
fix: removing `docker-run.sh` script

#407

CMCDragonkai commented on .gitlab-ci.yml:
```diff
@@ -321,16 +321,22 @@ integration:docker:
+        PK_TEST_DOCKER_IMAGE=$image \
+        PK_TEST_COMMAND=docker \
+        PK_TEST_PLATFORM=docker \
+        PK_TEST_DATA_DIR=/builds/$CI_PROJECT_PATH/tmp \
```
I don't think the additional env variable for `PK_TEST_DOCKER_IMAGE` as it adds more configuration.

I propose something like this:

```
PK_TEST_PLATFORM='docker'
PK_TEST_COMMAND="$(which docker) run \${options} someimage:sometag
```

The `someimage:sometag` is specified outside.

Now you need to use a simple utility function to "template" the options and args.

```
const template = (str, vars) => str.replace(
  /\${(.*?)}/g, 
  (_, p) => vars[p]
);

template(pkTestCommand, { options: '--x y z' })
```

CMCDragonkai replied:
That way the `PK_TEST_COMMAND` remains an actual command, it's just that options can be provided in an interpolated fashion.

And this allows future applications of `PK_TEST_COMMAND` to the other platforms in case there are special options required too.

Now where should the options be "templated"? It should be done in your `tests/utils/exec.ts`, you can create a common function to do it. The `template` can be a useful utility function put into `tests/utils/utils.ts`. Might even be useful elsewhere too.

CMCDragonkai replied:
You would need to make sure you're using `exec` and not `execFileSync`.

tegefaulkes replied:
I don't think that's going to work. when using `child_process.spawn` it takes (command, args, options). So the command will need to be `docker` on it's own, and the args need to be `['run', ...options, 'someimage:sometag', ...pkargs]`.

If we really need to get rid of the extra env variable, I can get the image id inside the test by execing `docker image ls -f "reference=polykey*" -q`

CMCDragonkai replied:
Doesn't the spawn take an arbitrary string as the first parameter? You should be able to pass a full command into `spawn`. You just need to enable `shell: true`.

CMCDragonkai replied:
![spawn](https://user-images.githubusercontent.com/640797/180921312-8ad0040b-4114-46e9-a9b2-9aa8dfec78fe.png)

CMCDragonkai replied:
So it should work like `child_process.spawn('ls -l | cat', [], { shell: true });`

CMCDragonkai replied:
BTW I just noticed `killSignal` is set to `SIGTERM`, that doesn't work on windows right? Seems strange that it would default to signal that has no effect on windows. Or maybe the docs is not up to date.

CMCDragonkai replied:
Ok so some new ideas since we discussed:

1. Try using `execFile` and `spawn` with `shell: true` option, and see if it's possible to pass an arbitrary command as the first argument. And what the second argument of `args` is used for, possibly just concatenated arguments after the string?
2. This allows us to use `PK_TEST_COMMAND='ls -l | docker run'` or any arbitrary shell command string. Do note that on windows, this will be a `CMD` shell and not a bash/sh shell.
3. The `PK_TEST_COMMAND` originally was just a path to a file, now its a shell command. If it's a shell command, it should support the ability to interpolate a shell variable like: `PK_TEST_COMMAND='docker run $DOCKER_OPTIONS image:tag'`. Where you can then interpolate `$DOCKER_OPTIONS` during `execFile` or `spawn` by using the env option such as `{ env: { ...process.env, DOCKER_OPTIONS: '--a --b --c' } }`.
4. The `PK_TEST_PLATFORM` and `PK_TEST_COMMAND` should be coupled together as all or nothing. Meaning they can only either be both set, or both unset (`undefined`). This check should be done on the `jest.config.js`. This ensures that the user has to set both, the platform for how it is run, and the command for what is run. This means that `PK_TEST_PLATFORM` should not be defaulted to `os.platform()` instead it defaults to undefined.
5. The end result is that target based tests will end up with an intermediate shell program, where as regular testing without the 2 env variables will use the original way of executing PK. We have to be wary of any kill signals now, since for docker there could be 2 intermediate processes the shell and the docker's shell (if using a shell), but if doing a direct execution, then it should be just the shell itself.
6. As for the `PK_TEST_DATA_DIRECTORY`, that should be renamed to `PK_TEST_TMPDIR` so it's obvious that it's overwriting the `os.tmpdir()` being used bby the tests. But it does not overwrite `TMPDIR` normally, as we don't want to change the meaning of the `TMPDIR`.

tegefaulkes replied:
Doing some testing with `child_process.spawn`. Looks like 1, 2, 3 above work as expected and should be doable. I'll make the changes and do some more testing.

tegefaulkes replied:
Just ran all of the tests locally. It works with 1 minor difference breaking a single notification test. When running with `shell: false` a sing arg in the args array can contain spaces and be considered a single arg. In the problem test we have the following args;

```ts
        [
          'notifications',
          'send',
          nodesUtils.encodeNodeId(receiverId),
          'test message 3',
        ],
```

With `shell: false` the arg `test message 3` is considered a single arg. as if we passed it to the command like `"test message 3"`. However  with `shell: true` each word is treated as a seperate arg. so the command things the message is just `test`.

This can be fixed by setting the message to `'"test message 3"'` but this will break the test when it runs normally. Since it's a pretty niche problem affecting only 1 test I'll just replace the spaces in the message with `_`.

Just keep this difference in behaviour in mind for the future.

tegefaulkes replied:
tracking what's been done here,
- [X] 1.
- [X] 2.
- [X] 3.
- [x] 4.
- [x] 5.
- [X] 6.

tegefaulkes replied:
For 4, we have 1 case where we can't set both. That's for normal testing in the CI. This means we'd either have to;
1. allow PK_TEST_PLATFORM to be set to `linux` with  `PK_TEST_COMMAND` being undefined.
2. Have `runTestifPlatform` run by default if `PK_TEST_PLATFORM` is not set.

I'm leaning towards 2. here but I may need to remove `'linux'` from every `runTestIfPlatform` usage.

CMCDragonkai replied:
What case is that? I was thinking that `PK_TEST_PLATFORM` is only useful during targeted execution. Otherwise tests should work normally across all platforms.

tegefaulkes replied:
Resolving this. It has been addressed.

tegefaulkes committed:
feat: `pkXTarget` works with any shell command

`DOCKER_OPTIONS` env variable containing all the docker run parameters to make the test work is provided to the command.

#410

tegefaulkes committed:
fix: changing `PK_TEST_DATA_DIR` to `PK_TEST_TMPDIR`

tegefaulkes committed:
fix: `PK_TEST_COMMAND` and PK_TEST_PLATFORM` must be set at the same time

The `runTestIfPlatform` commands will default to running if `PK_TEST_PLATFORM` is not set.

#410

tegefaulkes committed:
fix: updating nat tests to use `runDescribeIf`

tegefaulkes committed:
fix: consolidated all spawning functions in `tests/utils/exec.ts`

This is to key every major spawning command in one place in case we need to modify how we spawn using `ts-node`.

tegefaulkes committed:
fix: escaping args when using `shell: true`

tegefaulkes committed:
fix: `pkExecTarget` now uses `child_process.spawn`

Is is due to a compatability problems when using `shell: true` with `child_process.execFile`.
